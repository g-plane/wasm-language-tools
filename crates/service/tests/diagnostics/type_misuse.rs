use super::*;
use insta::assert_json_snapshot;
use wat_service::LanguageService;

#[test]
fn structs() {
    let uri = "untitled:test".to_string();
    let source = "
(module
  (type $func (func))
  (type $struct (struct (field (mut i32))))
  (type $array (array i32))
  (func (param (ref $struct))
    local.get 0
    local.get 0
    struct.get $struct 0
    struct.set $struct 0)
  (func (param (ref $struct))
    local.get 0
    local.get 0
    struct.get $func 0
    struct.set $array 0))
";
    let mut service = LanguageService::default();
    service.commit(uri.clone(), source.into());
    calm(&mut service, uri.clone());
    let response = service.pull_diagnostics(create_params(uri));
    assert_json_snapshot!(response);
}

#[test]
fn array() {
    let uri = "untitled:test".to_string();
    let source = "
(module
  (type $func (func))
  (type $struct (struct (field i32)))
  (type $array (array (mut i32)))
  (func (param (ref $array))
    local.get 0
    local.get 0
    i32.const 0
    array.get $array
    i32.const 0
    array.set $array)
  (func (param (ref $array))
    local.get 0
    local.get 0
    i32.const 0
    array.get $func
    i32.const 0
    array.set $struct))
";
    let mut service = LanguageService::default();
    service.commit(uri.clone(), source.into());
    calm(&mut service, uri.clone());
    let response = service.pull_diagnostics(create_params(uri));
    assert_json_snapshot!(response);
}

#[test]
fn array_copy() {
    let uri = "untitled:test".to_string();
    let source = "
(module
  (type $func (func))
  (type $struct (struct))
  (type $dst_array (array (mut i32)))
  (type $src_array (array i64))
  (func (param (ref $dst_array) (ref $src_array))
    local.get 0
    i32.const 0
    local.get 1
    i32.const 0
    i32.const 0
    array.copy $struct $struct)
  (func (param (ref $dst_array) (ref $src_array))
    local.get 0
    i32.const 0
    local.get 1
    i32.const 0
    i32.const 0
    array.copy $func $func)
  (func (param (ref $dst_array) (ref $src_array))
    local.get 0
    i32.const 0
    local.get 1
    i32.const 0
    i32.const 0
    array.copy $dst_array $src_array))
";
    let mut service = LanguageService::default();
    service.commit(uri.clone(), source.into());
    calm(&mut service, uri.clone());
    let response = service.pull_diagnostics(create_params(uri));
    assert_json_snapshot!(response);
}

#[test]
fn func() {
    let uri = "untitled:test".to_string();
    let source = "
(module
  (type $func (func))
  (type $struct (struct (field i32)))
  (type $array (array (mut i32)))
  (func (param (ref $func))
    local.get 0
    call_ref $func
    local.get 0
    return_call_ref $func)
  (func (param (ref $func))
    local.get 0
    call_ref $struct
    local.get 0
    return_call_ref $array))
";
    let mut service = LanguageService::default();
    service.commit(uri.clone(), source.into());
    calm(&mut service, uri.clone());
    let response = service.pull_diagnostics(create_params(uri));
    assert_json_snapshot!(response);
}
