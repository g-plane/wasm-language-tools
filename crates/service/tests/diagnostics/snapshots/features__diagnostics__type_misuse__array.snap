---
source: crates/service/tests/diagnostics/type_misuse.rs
expression: response
---
{
  "kind": "full",
  "items": [
    {
      "range": {
        "start": {
          "line": 16,
          "character": 4
        },
        "end": {
          "line": 16,
          "character": 19
        }
      },
      "severity": 1,
      "code": "type-check",
      "source": "wat",
      "message": "expected types [(ref null $func), i32], found [(ref $array), i32]",
      "relatedInformation": [
        {
          "location": {
            "uri": "untitled:test",
            "range": {
              "start": {
                "line": 14,
                "character": 4
              },
              "end": {
                "line": 14,
                "character": 15
              }
            }
          },
          "message": "expected type `(ref null $func)`, found `(ref $array)`"
        }
      ]
    },
    {
      "range": {
        "start": {
          "line": 18,
          "character": 4
        },
        "end": {
          "line": 18,
          "character": 21
        }
      },
      "severity": 1,
      "code": "type-check",
      "source": "wat",
      "message": "expected types [(ref null $struct), i32, any], found [(ref $array), any, i32]",
      "relatedInformation": [
        {
          "location": {
            "uri": "untitled:test",
            "range": {
              "start": {
                "line": 13,
                "character": 4
              },
              "end": {
                "line": 13,
                "character": 15
              }
            }
          },
          "message": "expected type `(ref null $struct)`, found `(ref $array)`"
        }
      ]
    },
    {
      "range": {
        "start": {
          "line": 16,
          "character": 14
        },
        "end": {
          "line": 16,
          "character": 19
        }
      },
      "severity": 1,
      "code": "type-misuse",
      "source": "wat",
      "message": "expected type is array, but type of `$func` is func",
      "relatedInformation": [
        {
          "location": {
            "uri": "untitled:test",
            "range": {
              "start": {
                "line": 2,
                "character": 2
              },
              "end": {
                "line": 2,
                "character": 21
              }
            }
          },
          "message": "func type defined here"
        }
      ]
    },
    {
      "range": {
        "start": {
          "line": 18,
          "character": 14
        },
        "end": {
          "line": 18,
          "character": 21
        }
      },
      "severity": 1,
      "code": "type-misuse",
      "source": "wat",
      "message": "expected type is array, but type of `$struct` is struct",
      "relatedInformation": [
        {
          "location": {
            "uri": "untitled:test",
            "range": {
              "start": {
                "line": 3,
                "character": 2
              },
              "end": {
                "line": 3,
                "character": 37
              }
            }
          },
          "message": "struct type defined here"
        }
      ]
    }
  ]
}
